text/0195-kubernetes-resource-detection.md

```diff
@@ -0,0 +1,120 @@
+# Environment Variable Resource Detection in Kubernetes
+
+This OTEP proposes to standardize environment variables used to detect resources exposed through the [Kubernetes downward API](https://kubernetes.io/docs/tasks/inject-data-application/environment-variable-expose-pod-information/) to provide a consistent experience on Kubernetes across languages and the collector.
+
+There is a mismatch between the format otel expects resources to be specified in for environment variables (a single, comma-deliniated list of key-value pairs in `OTEL_RESOURCE_ATTRIBUTES`), and the format Kubernetes exposes resource information in (one environment-variable for each attribute).  This makes correct resource detection in Kubernetes difficult, and the lack of specification has led to differing implementations across language SDKs and the collector.
+
+## Background
+
+Adding Kubernetes resources attributes to telemetry can be done one of three ways, depending on your telemetry collection pipeline:
+
+1. The application detects its own resource attributes from environment variables
+2. A "sidecar" collector container detects its own resource attributes from environment variables
+3. A centralized collector detects resource attributes for many Kubernetes workloads by querying the kube-apiserver.
+
+Method 3 is already implemented today as the `k8sattributesprocessor`, but requires running a collector, and adds additional load on the kube-apiserver. Methods 1 and 2 require substantial configuration today, which differs based on the language used, or if using a sidecar collector.
+
+Not all attributes in the [k8s semantic conventions](https://github.com/open-telemetry/opentelemetry-specification/blob/main/semantic_conventions/resource/k8s.yaml) are supported by the Kubernetes downward API.  The OpenTelemetry attributes that could be supported ([api reference](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.23/#envvarsource-v1-core)) are:
+
+* `k8s.pod.name`
+* `k8s.pod.uid`
+* `k8s.namespace.name`
+* `k8s.node.name`
+
+The current state of detecting downward-api-based resources differs, depending on where detection is being done:
+
+* The collector's [`resourcedetectionprocessor`](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/processor/resourcedetectionprocessor#resource-detection-processor) only accepts `OTEL_RESOURCE_ATTRIBUTES` for env-var-based resource detection, which does not work with the downward API.
```

Comment:
`OTEL_RESOURCE_ATTRIBUTES` can be set to `$(VAR_NAME)` which can substituted to the value from the downward API e.g.:

`OTEL_RESOURCE_ATTRIBUTES=k8s.container.name=myapp,k8s.pod.name=$(POD_NAME)`


See https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.22/#envvar-v1-core

Reply:
Updated the wording of this section, as it was outdated.  See the "alternative: using dependent environment variables" section below.